### ISSUE TITLE
10$: Warn on duplicate cases in switch

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#744]
```js
switch (a) {
  case "one":
    break;
  case "one":  // warning
    break;
  case "three":
    break;
}
```

**Comment by Rebraws** [comment]
Is using a `hash_set<string8_view>` a good approach for this issue?

I was thinking about using the hash_set and then in this else case https://github.com/quick-lint/quick-lint-js/blob/0f9d8c623cd5a8fc5c2848deb53641f49defdddd/src/quick-lint-js/fe/parse-statement.cpp#L1853 do something like this:

```
} else {  
        string8_view case_expression = this->peek().span().string_view();
        if (the_hash_set.contains(case_expression)) {
            // GIVE THE WARNING
         }
        the_hash_set.insert(case_expression);
        this->parse_and_visit_expression(
            v, precedence{
                   .colon_type_annotation = allow_type_annotations::never,
               });
        QLJS_PARSER_UNIMPLEMENTED_IF_NOT_TOKEN(token_type::colon);
        this->skip();
      }
```

**Comment by strager** [comment]
> `string8_view case_expression = this->peek().span().string_view();`

This looks wrong. You're only comparing the token after `case`, but there might be multiple tokens. For example, this should not report a warning:
```javascript
switch (x) {
case MyEnum.A:
case MyEnum.B:
case MyEnum.C:
  // ...
}
```

I think we should instead call `parse_expression` and `visit_expression` separately (instead of `parse_and_visit_expression`), then use the span of the expression.

> Is using a `hash_set<string8_view>` a good approach for this issue?

I think so, assuming the string8_view refers to the span of the entire expression. That would work with string literals, number literals, and constant references.

Your idea won't work if the case expressions are spelled slightly differently. But because exact-copy-paste errors are more common than things like `case "x": case "\x78":`, I think comparing string spans is the way to go.

**Comment by strager** [comment]
Did you want to be assigned this task? It looks like you're half way to finishing it already. =]
